### PR TITLE
Instrument event dispatcher (rebased)

### PR DESCRIPTION
**Pull request by chancancode**
Rebased version of #12088

### CODE REVIEW

File: packages/ember-routing/lib/system/router.js [hunk 1]
```diff
@@ -542,6 +553,8 @@ var EmberRouter = EmberObject.extend(Evented, {
       emberRouter.didTransition(infos);
     };
 
+    // TODO
+    //router.willTransition = this.willTransition.bind(this);
```
Comment (rwjblue) [inline packages/ember-routing/lib/system/router.js]: What is this comment about?

Comment (chancancode) [inline packages/ember-routing/lib/system/router.js]: Hm, I think this was us saying we should just replace the line below with a `bind` since that's exactly what it was doing, but maybe we didn't want to change unrelated code in the commit

**Comment by rwjblue** [comment]
Looks like this needs a rebase.

**Comment by mixonic** [comment]
Seems inoffensive, but does lack any tests for the action and routing instrumentation.

**Comment by rwjblue** [comment]
There is a very long trail on this one:

* #12088 
* https://github.com/emberjs/ember.js/pull/12550
* https://github.com/emberjs/ember.js/pull/13017
* https://github.com/emberjs/ember.js/pull/13279

Closing this version for now...
